### PR TITLE
[Performance][SFA] Shard LightningIndexer across DCP ranks

### PR DESCRIPTION
**Pull request by Ruiqiu-Zheng**
## Summary

This draft adds an opt-in **SFA DCP-sharded LightningIndexer** path for long-context prefill.

The existing replicated path remains the default and fail-closed fallback. The sharded path keeps **all query rows on every DCP rank**, shards only the LightningIndexer **K-domain/cache ownership**, writes rank-local indexer cache entries, remaps local indices back to global indices, and recovers the global TopK across DCP16 with a 4-round XOR butterfly merge.

LightningIndexer selection remains owned by `AscendSFAIndexerBackend`; the SFA consumer remains a consumer and does not take over LI selection.

## Why this is one change

The selector, KV allocation factor, rank-local cache writes, DCP-local metadata geometry, and distributed TopK merge form one physical-cache correctness unit:

```text
default-off selector
        ↓
indexer cache allocation: replicated factor 16 → rank-local factor 1
        ↓
DCP-local K/cache ownership + local LI selection with scores
        ↓
local index → global index remap
        ↓
4-round DCP16 butterfly TopK merge
        ↓
existing SFA consumer receives global indices
```

Changing the allocation factor without the rank-local write/select/merge path would be unsafe, while the algorithm without the selector/allocation plumbing would be dead code. This PR therefore keeps those pieces together and keeps unsupported configurations on the incumbent replicated path.

## Supported envelope

The opt-in sharded path is intentionally narrow:

- TP16 / DCP16
- CP KV-cache interleave = 128
- PCP = 1
- eager mode
- NPU/cudagraph disabled
- DSA-CP disabled
- sparse LI C8 disabled
- xLite graph disabled
- feature default-off
- **no KV transfer / PD connector configured**; current Mooncake/SFA transfer still assumes the replicated indexer-K cache layout

Unsupported topology/runtime/configuration, including any non-`None` `kv_transfer_config`, falls back to the incumbent replicated indexer.

There is **no hard-coded context-length profitability threshold**. The feature remains default-off; the measurements below document the currently validated region rather than pretending two context points define a universal crossover.

## Performance and capacity evidence

All claims below are bounded to the tested lab lineage and do not constitute production validation.

| Protocol | Baseline | Candidate | Delta |
|---|---:|---:|---:|
| C64K/Q4K C1 TTFT p50 | 1848.21 ms | 1803.18 ms | **-45.03 ms / -2.44%** |
| C64K/Q4K C1 TTFT p90 | 1863.88 ms | 1807.15 ms | **-56.73 ms / -3.04%** |
| C96K/Q4K C1 TTFT p50 | 2125.41 ms | 1889.03 ms | **-236.38 ms / -11.12%** |
| C96K/Q4K C1 TTFT p90 | 2136.29 ms | 1893.46 ms | **-242.83 ms / -11.37%** |
| C1 KV capacity | 208,896 tokens | 385,024 tokens | **+176,128 / +84.31%** |

C64K and C96K matched runs passed the bounded correctness checks used for the PR claim.

C128K/C192K showed stronger latency scaling, but those points exposed incumbent-level output nondeterminism across exact prompt/cache-reset repeats. They are therefore treated only as scaling observations and are **not** correctness-PASS claims.

### C96K/Q4K, concurrency 2 on one TP16/DCP16 worker

With `gpu_memory_utilization=0.82`, `max_num_seqs=2`, `max_num_batched_tokens=8192`:

- incumbent replicated capacity: **120,832** tokens, below two 98,304-token prefixes (196,608 required)
- sharded candidate capacity: **249,856** tokens
- capacity delta: **+129,024 / +106.78%**
- 8/8 timed candidate requests retained exactly **98,304 cached tokens**
- 16-rank consistency passed; no corruption or cache shortfall
- candidate p50 executed-Q throughput: **2187.01 tok/s**

There is no matched incumbent C2 speedup claim because the incumbent cannot maintain both full prefixes in this exact geometry.

### Tested concurrency-3 boundary

With `max_num_seqs=3` and `max_num_batched_tokens=12288`, candidate KV capacity was **112,640** tokens, below the 294,912-token three-prefix requirement. The run therefore stopped at the capacity decision instead of producing a misleading timing number.

This establishes only the tested single-worker boundary; it is not a claim that every possible C3 configuration is impossible.

### Two independent TP16/DCP16 Prefill workers

A bounded horizontal-replication canary ran one C96/Q4K C2 workload on server09+10 and another independently on server11+12:

- both workers: sharded selector enabled, cache factor 1
- KV capacity: 247,808 / 249,856 tokens
- 12/12 timed requests retained exactly 98,304 cached tokens
- both workers: 16-rank consistency PASS, zero fatal errors, zero cache shortfalls
- p50 Q throughput: **2184.45 + 2188.15 = 4372.59 tok/s**
- sum of worker p50 throughput is **99.97%** of the cross-run 2× single-worker reference

This is evidence that no material cross-worker compute collapse was observed. It is **not** a production 2× scaling claim; the one-shot union-window metric included start skew/synchronization overhead and is not used as a production scaling number.

## Correctness contract

The sharded algorithm keeps all Q rows and shards only K/cache ownership:

1. each rank computes its local visible K count;
2. LI runs on the rank-local K/cache view and returns local indices plus selected scores;
3. local indices are remapped to global indices;
4. DCP16 performs 4 pairwise XOR merge rounds (`xor 1/2/4/8`), retaining TopK after each round;
5. unique cutoffs are exact; cutoff ties preserve the established threshold-equivalence contract and repeat determinism;
6. inactive local candidates use sentinel index `-1` / score `-inf` semantics.

Current bounded validation includes native DCP16 correctness, long-context C64/C96 full-model correctness, repeat/tie coverage, metadata/cache mapping tests, fallback tests, and the C2 serviceability checks above.

## Review hardening already addressed

- **P2P peer rank:** checked against the pinned PyTorch 2.10.0 / torch-npu 2.10.0.post4 semantics. `P2POp.peer` is supplied as the global peer and the process-group path derives the group-relative peer; the existing `dcp_group.ranks[peer_in_group]` usage is therefore intentional.
- **metadata host-device synchronization:** latest commit avoids NPU tensor `.item()` for max query/sequence lengths when CPU metadata is available and preserves the existing device fallback when it is not.

## CI / current-main status

Current PR head: `66477f207210fcf17895c578316e389152d6a87b`

- DCO: PASS
- pre-commit: PASS
- CPU UT: PASS
- selected tests: waiting for maintainer `ready-precise`
- `ci-gate`: currently fails only because the selected-test trigger label is absent
- latest inspected upstream main: `12be3d34a1510be8e19542f577e732a12c47a1d1`
- merge-tree: 0 textual conflicts
- no new drift in the candidate-overlap files since the previous semantic review

## Non-claims

This PR does **not** claim:

- production validation or production gain;
- sharded-indexer operation with Mooncake or any other KV-transfer connector; current KV-transfer configurations deliberately fail closed to the replicated incumbent until connector-side sharded indexer-K transfer semantics exist;
- graph-mode support;
- PCP>1, DSA-CP, sparse-LI-C8, xLite graph, TP32/DCP32, or mixed prefill+decode support;
- C128K/C192K correctness PASS;
- a universal context-length break-even threshold;
- matched incumbent C2 speedup;
- production-linear horizontal scaling.

The intended next step is the repository-selected CI on the current head, followed by human review. No additional feature geometry is planned for this PR.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

**Comment by Ruiqiu-Zheng** [comment]
Hi @weijinqian0 @zzzzwwjj @realliujiaxu — the latest head `7b4c6745bba65780616b491b745db14eac9b2fbb` now has pre-commit and CPU UT passing. I also addressed the valid Gemini feedback by avoiding NPU tensor `.item()` synchronization in the SFA DCP metadata hot path when CPU metadata is available, while preserving the existing fallback. The P2P peer-rank concern was checked against the pinned PyTorch 2.10.0 / torch-npu 2.10.0.post4 semantics, where the current global-peer usage is expected. Could one of you please add the `ready-precise` label when convenient so the selected tests can run? Thanks.

**Comment by Ruiqiu-Zheng** [comment]
@codex review

Please re-review the current head `66477f207210fcf17895c578316e389152d6a87b`. The KV-transfer cache-layout P1 from the previous review is addressed by failing closed to the replicated incumbent whenever `kv_transfer_config` is configured, with focused selector/cache-factor regression coverage. Please evaluate the current head, including whether any correctness blocker remains.

File: vllm_ascend/attention/indexer.py [hunk 1]
```diff
@@ -183,6 +311,130 @@ def __init__(self, vllm_indexer: nn.Module, qk_rope_head_dim: int) -> None:
...
+                    tag=321 + level * 2,
+                ),
+                dist.P2POp(
+                    dist.irecv,
+                    other_indices,
+                    peer,
+                    group=dcp_group.device_group,
+                    tag=321 + level * 2,
+                ),
+            ]
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/indexer.py:461]: Checked against the pinned PyTorch v2.10.0 implementation. `P2POp(peer=..., group=...)` treats `peer` as the global rank and canonicalizes it to `group_peer`; `batch_isend_irecv` then passes that group-relative value as `group_dst/group_src`. The current `peer = dcp_group.ranks[peer_in_group]` is therefore intentional. Passing `peer_in_group` through the `peer=` argument would be interpreted as a global rank and can select the wrong process when the DCP group is not `[0..world_size-1]`.

File: vllm_ascend/attention/indexer.py [hunk 2]
```diff
@@ -512,16 +945,117 @@ def build(
...
+            local_positions = input_positions[dcp_local_token_mask]
+            local_block_cols = max(
+                1,
+                int(
+                    torch.ceil(
+                        common_attn_metadata.seq_lens[:num_reqs].max().float()
+                        / float(self.kernel_block_size * self._dcp_world_size)
+                    ).item()
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/indexer.py:1020]: Addressed in the current head. The DCP-sharded metadata path now uses `_get_dcp_metadata_max_lens_from_cpu()` when `query_start_loc_cpu` and `seq_lens_cpu_upper_bound` are available, so the validated long-prefill hot path avoids the NPU `.item()` syncs. The device `.item()` path remains only as a compatibility fallback when CPU-authoritative metadata is unavailable.

File: vllm_ascend/utils.py [hunk 3]
```diff
@@ -152,6 +152,85 @@ def enable_sfa_dcp_replicated_indexer(vllm_config: VllmConfig | None = None) ->
...
+    ):
+        return False
+
+    compilation_config = getattr(vllm_config, "compilation_config", None)
+    cudagraph_mode = getattr(compilation_config, "cudagraph_mode", None)
+    if cudagraph_mode is not None:
+        mode_name = getattr(cudagraph_mode, "name", str(cudagraph_mode).rsplit(".", 1)[-1])
+        if str(mode_name).upper() != "NONE":
+            return False
+    return True
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/utils.py:219]: Addressed in the current head. `enable_sfa_dcp_sharded_indexer()` now returns `False` whenever `vllm_config.kv_transfer_config` is non-None, before the cache factor is selected. KV-transfer configurations therefore stay on the incumbent replicated-indexer layout.

File: vllm_ascend/attention/indexer.py [hunk 4]
```diff
@@ -183,6 +324,124 @@ def __init__(self, vllm_indexer: nn.Module, qk_rope_head_dim: int) -> None:
...
+            raise RuntimeError("DCP sharded indexer butterfly merge is restricted to DCP16.")
+        rank_in_group = dcp_group.rank_in_group
+        indices = global_indices.contiguous()
+        scores = local_scores.contiguous()
+        # Four pairwise merge rounds reduce communication from publishing all
+        # 16 rank-local K lists to every rank at once to O(K log DCP).  Both
+        # peers order lower-rank subgroup first, making cutoff-tie selection
+        # repeat-deterministic while preserving the established threshold
+        # equivalence contract.
+        for level, step in enumerate((1, 2, 4, 8)):
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/indexer.py:427]: Addressed in the current PR scope. The head now includes `tests/e2e/nightly/single_node/models/test_sfa_dcp_sharded_indexer_perf.py`, an A3 TP16/DCP16 eager C96K/Q4K matched replicated-vs-sharded TTFT regression test with a bounded regression threshold. The selected NPU matrix still requires the maintainer CI trigger label.

File: vllm_ascend/attention/indexer.py [hunk 5]
```diff
@@ -512,16 +920,116 @@ def build(
...
+            local_indices = (
+                local_positions // (self._dcp_interleave_size * self._dcp_world_size) * self._dcp_interleave_size
+            )
+            local_indices = local_indices + local_positions % self._dcp_interleave_size
+            local_block_idx = local_indices // self.kernel_block_size
+            local_block_offsets = local_indices % self.kernel_block_size
+            dcp_local_slot_mapping = (
+                dcp_local_request_block_table[local_req_indices, local_block_idx.to(torch.long)]
+                * self.kernel_block_size
+                + local_block_offsets
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/indexer.py:1042]: Addressed in the current head. The rank-local write mask is now intersected with `arange(num_input_tokens) < num_actual_tokens` before `local_positions`, `local_req_indices`, and `dcp_local_slot_mapping` are formed. Padded tokens therefore retain invalid/no-write semantics and cannot be remapped into block 0.
